src/shims/time.rs

```diff
@@ -1,8 +1,71 @@
+use std::sync::atomic::AtomicU64;
+use std::sync::Arc;
 use std::time::{Duration, Instant, SystemTime};
 
+use rustc_data_structures::sync::Ordering;
+
 use crate::concurrency::thread::Time;
 use crate::*;
 
+#[derive(Debug)]
+pub enum TimeAnchor {
+    Host(Instant),
+    Virtual(Arc<AtomicU64>),
+}
+
+impl TimeAnchor {
+    pub fn new(communicate: bool) -> Self {
+        if communicate { Self::Host(Instant::now()) } else { Self::Virtual(Arc::new(0.into())) }
+    }
+
+    pub fn get(&self) -> Duration {
+        match self {
+            Self::Host(instant) => Instant::now().saturating_duration_since(*instant),
+            Self::Virtual(nanoseconds) => Duration::from_nanos(nanoseconds.load(Ordering::Relaxed)),
+        }
+    }
+
+    pub fn tick(&self) {
+        match self {
+            Self::Host(_) => (),
+            Self::Virtual(nanoseconds) => {
+                nanoseconds.fetch_add(1, Ordering::Relaxed);
+            }
+        }
+    }
+
+    pub fn sleep(&self, duration: Duration) {
+        match self {
+            Self::Host(_) => std::thread::sleep(duration),
+            Self::Virtual(nanoseconds) => {
+                nanoseconds.fetch_add(duration.as_nanos().try_into().unwrap(), Ordering::Relaxed);
```

Comment:
```suggestion
                // Just pretend that we have slept for some time.
                nanoseconds.fetch_add(duration.as_nanos().try_into().unwrap(), Ordering::Relaxed);
```

Reply:
Is this a joke about our sleeping habits?